### PR TITLE
[PeerTube] Add support for PeerTube v6 features

### PR DESCRIPTION
**Pull request by TobiGr**
- [x] I carefully read the [contribution guidelines](https://github.com/TeamNewPipe/NewPipe/blob/HEAD/.github/CONTRIBUTING.md) and agree to them.
- [x] I have tested the API against [NewPipe](https://github.com/TeamNewPipe/NewPipe).
- [x] I agree to create a pull request for [NewPipe](https://github.com/TeamNewPipe/NewPipe) as soon as possible to make it compatible with the changed API.

# Description

[PeerTube 6.0.0 ](https://joinpeertube.org/news/release-6.0) added support for multiple new features which are implemented in this PR:

- [x] support for stream segments/chapters extraction: implement `PeerTubeStreamExtractor.getStreamSegments()`
- [x] support for storyboard / video preview: implement `PeerTubeStreamExtractor.getFrames()`

# Videos for testing

- https://tube.tchncs.de/w/6U8A1AZ8nEkzgZXwf2kdW7
- https://framatube.org/w/vqABGP97fEjo7RhPuDnSZk

### CODE REVIEW

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java [hunk 1]
```diff
@@ -316,6 +318,58 @@ public String getSupportInfo() {
...
+    @Nonnull
+    @Override
+    public List<StreamSegment> getStreamSegments()
+            throws ParsingException, IOException, ReCaptchaException {
+        final List<StreamSegment> segments = new ArrayList<>();
+        final JsonObject segmentsJson = fetchSubApiContent("chapters");
+        if (segmentsJson != null && segmentsJson.has("chapters")) {
+            final JsonArray segmentsArray = segmentsJson.getArray("chapters");
+            for (final Object segment : segmentsArray) {
+                if (segment instanceof JsonObject) {
```
Comment (Profpatsch) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:330]: why not segment.getObject()? Is the lenient parsing intentional?

**Comment by AudricV** [comment]
I don't think you need to do breaking changes. Can't you catch errors when getting thumbnails and chapters like done for description and throwing an `ExtractionException` wrapping the original one?

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java [hunk 2]
```diff
@@ -316,6 +318,65 @@ public String getSupportInfo() {
...
+                    final int height = storyboardObject.getInt("spriteHeight");
+                    final int totalWidth = storyboardObject.getInt("totalWidth");
+                    final int totalHeight = storyboardObject.getInt("totalHeight");
+                    final int framesPerPageX = totalWidth / width;
+                    final int framesPerPageY = totalHeight / height;
+                    final int count = framesPerPageX * framesPerPageY;
+                    final int durationPerFrame = storyboardObject.getInt("spriteDuration") * 1000;
+
+                    framesets.add(new Frameset(
+                            List.of(baseUrl + url), // there is only one image per video
```
Comment (Profpatsch) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:370]: One per video? in https://joinpeertube.org/news/release-6.0 it looks like there’s multiple per video

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java [hunk 3]
```diff
@@ -636,6 +697,41 @@ private void addNewVideoStream(@Nonnull final JsonObject streamJsonObject,
         }
     }
 
+    /**
+     * Fetch content from a sub-API of the video.
+     * @param subPath the API subpath after the video id,
+     *                e.g. "chapters" for "/api/v1/videos/{id}/storyboards"
```
Comment (Profpatsch) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:703]: "storyboards" for …

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java (hunk 2 above)
Comment (Profpatsch) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:370]: Ah, it’s a composite image containing all of the thumbnails e.g. https://tube.tchncs.de/lazy-static/storyboards/8b7510d4-04bd-4937-90fd-2714c8cb4887.jpg

**Review by Profpatsch — APPROVED** [review]
I tested it against current newpipe with the example videos and it seems to work pretty well.

**Comment by Profpatsch** [comment]
One bug I found: The timestamps in chapters are missing:

image

**Comment by Profpatsch** [comment]
It always jumps to the beginning when I select a chapter cause the timestamp is `0:00`.

**Review by Profpatsch — DISMISSED** [review]
check timestamps

File: extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java [hunk 4]
```diff
@@ -316,6 +318,65 @@ public String getSupportInfo() {
...
+        } catch (final IOException | ReCaptchaException e) {
+            throw new ParsingException("Could not get stream segments", e);
+        }
+        if (segmentsJson != null && segmentsJson.has("chapters")) {
+            final JsonArray segmentsArray = segmentsJson.getArray("chapters");
+            for (int i = 0; i < segmentsArray.size(); i++) {
+                final JsonObject segmentObject = segmentsArray.getObject(i);
+                segments.add(new StreamSegment(
+                        segmentObject.getString("title"),
+                        segmentObject.getInt("start")));
```
Comment (Profpatsch) [inline extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:337]: The json field is called `timecode`!

```
{
    "chapters": [
        {
            "timecode": 0,
            "title": "Vorgeschichte/Einleitung"
        },
        {
            "timecode": 57,
            "title": "CPU-Auswahl"
        },
        {
            "timecode": 204,
            "title": "Grundbeschaltung"
```

Comment (Profpatsch) [reply extractor/src/main/java/org/schabi/newpipe/extractor/services/peertube/extractors/PeertubeStreamExtractor.java:337]: Can confirm that fixes it

**Comment by Profpatsch** [comment]
In what way is that relevant?

**Review by Stypox — APPROVED** [review]
Looks good to me. I tested and it works. Thank you!

**Comment by Stypox** [comment]
I guess the reason is that I didn't notice that `yt-fix_comments_extraction` (i.e. the unmerged #1163) was 23 commits behind `dev`, otherwise I would have rebased that PR on top of dev.

**Comment by Stypox** [comment]
Yes, I will for sure for the next RC

**Pull request merged**